Microsoft.Advisor new api version 2024-11-18-preview

rita-lu opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

Commit:
Create baseline commit from 2023-09-01-preview to new 2024-11-18-preview version

Commit:
Update api version in files to new version 2024-11-18-preview

Commit:
Add new triageResources RT

Commit:
Delete Assessments swagger files

Commit:
Fix avocado errors

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[sdk/resourcemanager/advisor/armadvisor](https://apiview.dev/Assemblies/Review/31662be051d24dafb9f0b928061800c3?revisionId=9bda05495ca94be4a13b4faf51fc9c43)
[Microsoft.Advisor](https://apiview.dev/Assemblies/Review/5ef8b9dcef954e74afd9a6857288ead7?revisionId=98de19f91ebd4f3da653849746894e7f)

Commit:
fix prettier issue

rita-lu commented:
The API version has been approved in the private repo - https://github.com/Azure/azure-rest-api-specs-pr/pull/21753
We have removed the Assessments.json related Swagger files but there is a Swagger Avocado error "Please make sure the missing API swaggers are in the default tag." Not sure how to fix this as i've already removed assessments.json from the readme file

gary-x-li commented:
> The API version has been approved in the private repo - [Azure/azure-rest-api-specs-pr#21753](https://github.com/Azure/azure-rest-api-specs-pr/pull/21753) We have removed the Assessments.json related Swagger files but there is a Swagger Avocado error "Please make sure the missing API swaggers are in the default tag." Not sure how to fix this as i've already removed assessments.json from the readme file

Have you checked [all the details](https://github.com/Azure/avocado/blob/main/README.md#missing_apis_in_default_tag) regarding this error? If so and you still have trouble fixing it, please post a message in [this Teams channel](https://teams.microsoft.com/l/channel/19%3A0351f5f9404446e4b4fd4eaf2c27448d%40thread.skype/API%20Spec%20Review?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)

rita-lu commented:
We do not want to support the assessments.json for this specific version because it belongs to another team that is sharing the same RP as us and we do not need to add those endpoints for this version. There is no customer impact. Just that customers will not be able to use the assessments related endpoints for this new version.

gary-x-li commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,927 @@
+          "readOnly": true,
+          "description": "Type of resource this triageResource corresponds to e.g. \"Cosmos DB\"."
+        },
+        "resourceId": {
```
>resourceId [](http://example.com/codeflow?start=9&length=10)

Add `"format": "arm-id"`? Also consider adding x-ms-arm-id-details: [https://azure.github.io/autorest/extensions/#x-ms-arm-id-details](https://azure.github.io/autorest/extensions/#x-ms-arm-id-details)

gary-x-li commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,927 @@
+          "readOnly": true,
+          "description": "Unique identifier for the recommendation resource this triageResource belongs to."
+        },
+        "subscriptionId": {
```
>subscriptionId [](http://example.com/codeflow?start=9&length=14)

Add `"format": "uuid"`?

TimLovellSmith replied:
+1 that, are these uuids?

Commit:
Resolve arm review comments

Commit:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into ritalu/Microsoft.Advisor-2024-11-18-preview

Commit:
remove uuid

rkmanda commented:
> We do not want to support the assessments.json for this specific version because it belongs to another team that is sharing the same RP as us and we do not need to add those endpoints for this version. There is no customer impact. Just that customers will not be able to use the assessments related endpoints for this new version. Can the Avocado error be suppressed?

Rita, its not an option to not support older APIs in a new API version. Please get an exception from the Breaking change review board if you want to go ahead with this change. Please review the versioning policies here - https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#api-versioning

Commit:
Add back assessments endpoints

Commit:
Add back assessments endpoints

rita-lu commented:
> > We do not want to support the assessments.json for this specific version because it belongs to another team that is sharing the same RP as us and we do not need to add those endpoints for this version. There is no customer impact. Just that customers will not be able to use the assessments related endpoints for this new version. Can the Avocado error be suppressed?
> 
> Rita, its not an option to not support older APIs in a new API version. Please get an exception from the Breaking change review board if you want to go ahead with this change. Please review the versioning policies here - https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#api-versioning

I've added back the older APIs to this new version

Commit:
suppress lintdiff errors

TimLovellSmith commented on specification/advisor/resource-manager/readme.md:
```diff
@@ -26,7 +26,26 @@ These are the global settings for the Advisor API.
+    from:
+      - assessments.json
+  - code: RequiredPropertiesMissingInResourceModel
+    reason: Suppressing for now to avoid breaking change.
```
what error was this suppressing?

TimLovellSmith replied:
❌ RequiredPropertiesMissingInResourceModel Model definition 'AssessmentTypeListResult' must have the properties 'name', 'id' and 'type' in its hierarchy and these properties must be marked as readonly.
Location: Microsoft.Advisor/preview/2024-11-18-preview/assessments.json#L386 RPC-Get-V1-03, RPC-Put-V1-08
❌ RequiredPropertiesMissingInResourceModel Model definition 'WorkloadListResult' must have the properties 'name', 'id' and 'type' in its hierarchy and these properties must be marked as readonly.
Location: Microsoft.Advisor/preview/2024-11-18-preview/assessments.json#L429 RPC-Get-V1-03, RPC-Put-V1-08

TimLovellSmith replied:
This sounds suspiciously like an authoring error!!

TimLovellSmith replied:
This seems to be because AssessmentTypeListResult and WorkloadListResult lists do not follow the Azure RPC - their contained resource types are not proper resource declarations. I'm reluctant to say we should perpetuate this problem further. Yet we don't want to avoid breaking changes... 
expedient path forward: we should scope down these suppressions with constraints, to explicitly the least scope possible so we don't accidentally suppress other linter errors

rita-lu replied:
These issues were already present in the other API version that we have merged in the past. Are you saying that we need to fix this for the previous api version and this version?

TimLovellSmith replied:
@rita-lu Sorry, I need to be more clear: what I want you to do, is enhance the suppression annotations by making them only suppress the issue for this particular API path. That way, they will not suppress the issue if it happens on any new API path in the future.

TimLovellSmith replied:
(for suppression review approval)

rita-lu replied:
Have added scoped suppresssion

TimLovellSmith commented on specification/advisor/resource-manager/readme.md:
```diff
@@ -26,7 +26,26 @@ These are the global settings for the Advisor API.
+  - Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json
+suppressions:
+  - code: ArmResourcePropertiesBag
+    reason: Suppressing for now to avoid breaking change.
```
Looks like it was ❌ ArmResourcePropertiesBag Top level property names should not be repeated inside the properties bag for ARM resource 'AssessmentResult'. Properties [properties.type] conflict with ARM top level properties. Please rename these.
Location: Microsoft.Advisor/preview/2024-11-18-preview/assessments.json#L284 RPC-Get-V1-07

TimLovellSmith replied:
I'm confused about this still. Will need to look closer at AssessmentResult resources perhaps

TimLovellSmith replied:
@rita-lu here also, please enhance the suppression annotations - please add a path constraint to they only suppress the issue for this particular API path. That way, they will not suppress the issue if it happens on any new API path in the future.

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          "description": "Triage resource properties.",
+          "x-ms-client-flatten": true
+        }
+      }
```
Instead of doing this way, I'd suggest you use 'allOf' to inherit ProxyResource from common-types.

razvanbadea-msft replied:
@rita-lu check this comment

rita-lu replied:
Resolved

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          "type": "string",
+          "readOnly": true,
+          "description": "Type of resource this triageResource corresponds to e.g. \"Cosmos DB\"."
+        },
```
Possibly clarify in the description: is this meant to be a programmatic resource type identifier, or just a human readable one that might change over time?

rita-lu replied:
What is the purpose of including this? We do not want to include that in the description.

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+        "resourceId": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Full Azure Resource Id path of the resource this triageResource corresponds to.",
```
>Resource Id [](http://example.com/codeflow?start=37&length=11)

nit don't need to capitalize 'resource id' in description

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          "readOnly": true,
+          "description": "Unique identifier for the review resource this triageResource belongs to."
+        },
+        "recommendationId": {
```
>recommendationId [](http://example.com/codeflow?start=9&length=16)

consider naming it recommendationUniqueId

rita-lu replied:
We are already using this in a previous version and we do not want to change this now.

TimLovellSmith replied:
You don't have to take this one. This is not a mandatory fix. I just want to take the opportunity to make a suggestion on what might be more user friendly if this isn't yet shipped to customers.

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          "type": "string",
+          "readOnly": true,
+          "description": "Name of the resource group this triageResource belongs to."
+        },
```
Why include this when its already part of resourceId?

rita-lu replied:
We are already using this in a previous version and we do not want to change this now.

TimLovellSmith replied:
No need to break customers. #wontfix

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          "type": "string",
+          "readOnly": true,
+          "description": "Unique identifier for the subscription resource this triageResource belongs to."
+        },
```
Why include this? It seems like its already part of resourceId?

rita-lu replied:
We are already using this for customers and we do not want to change this now. The PR was approved in the private repo already. Why are we having new comments now?

TimLovellSmith replied:
So #wontfix right

TimLovellSmith commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,928 @@
+          }
+        },
+        "nextLink": {
+          "type": "string",
```
>   "type": "string", [](http://example.com/codeflow?start=7&length=20)

consider "format": "uri"

TimLovellSmith replied:
This one is basically optional too, its just a nicety of OpenAPI spec format.

Commit:
Resolve review comment

Commit:
Add scoped suppression

Commit:
Fix suppression

rita-lu commented:
/azp

rita-lu commented:
/azp run

msyyc commented:
/azp run

Commit:
resolve review comment

Commit:
fix space

razvanbadea-msft commented on specification/advisor/resource-manager/Microsoft.Advisor/preview/2024-11-18-preview/resiliencyReviews.json:
```diff
@@ -0,0 +1,913 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"
```
>v3 [](http://example.com/codeflow?start=71&length=2)

I see in all the api files a mix usage of v3/v4 for common-types. This is not recommended, can you update them to v5 or v6? Or take this into consideration for a future pr

rita-lu replied:
Yes we can do it in the next PR

rkmanda replied:
mixing versions of common types in the same open api spec json will cause issues when you try to generate the SDKs for these APIs later. Please make the version dependencies same across all references.. so all v3 or all v4.

rkmanda reviewed (approved):
:shipit:
